// scichart/Charting/Visuals/RenderableSeries/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["BRUSH_UP"] = "BRUSH_UP";
    PROPERTY["BRUSH_DOWN"] = "BRUSH_DOWN";
    PROPERTY["COLOR"] = "COLOR";
    PROPERTY["COLOR_MAP"] = "COLOR_MAP";
    PROPERTY["CONTOUR_DRAWING_MODE"] = "CONTOUR_DRAWING_MODE";
    PROPERTY["CONTOUR_MINORS_PER_MAJOR"] = "CONTOUR_MINORS_PER_MAJOR";
    PROPERTY["CONTOUR_MINOR_LINE_STYLE"] = "CONTOUR_MINOR_LINE_STYLE";
    PROPERTY["CONTOUR_MAJOR_LINE_STYLE"] = "CONTOUR_MAJOR_LINE_STYLE";
    PROPERTY["CONTOUR_COLOR_MAP_MODE"] = "CONTOUR_COLOR_MAP_MODE";
    PROPERTY["CORNER_RADIUS"] = "CORNER_RADIUS";
    PROPERTY["DATA_POINT_WIDTH"] = "dataPointWidth";
    PROPERTY["DATA_POINT_WIDTH_MODE"] = "DATA_POINT_WIDTH_MODE";
    PROPERTY["DATA_SERIES"] = "dataSeries";
    PROPERTY["DRAWING_PROVIDERS"] = "DRAWING_PROVIDERS";
    PROPERTY["DRAW_NAN_AS"] = "drawNaNAs";
    PROPERTY["DRAW_CONNECTOR"] = "DRAW_CONNECTOR";
    PROPERTY["DRAW_WHISKERS"] = "DRAW_WHISKERS";
    PROPERTY["EFFECT"] = "EFFECT";
    PROPERTY["ENABLE_DRAWING_OPTIMISATIONS"] = "ENABLE_DRAWING_OPTIMISATIONS";
    PROPERTY["ERROR_DIRECTION"] = "ERROR_DIRECTION";
    PROPERTY["ERROR_MODE"] = "ERROR_MODE";
    PROPERTY["FILL"] = "fill";
    PROPERTY["FILL_Y1"] = "FILL_Y1";
    PROPERTY["FILL_LINEAR_GRADIENT"] = "FILL_LINEAR_GRADIENT";
    PROPERTY["FILL_LINEAR_GRADIENT_Y1"] = "FILL_LINEAR_GRADIENT_Y1";
    PROPERTY["FILL_VALUES_OUT_OF_RANGE"] = "FILL_VALUES_OUT_OF_RANGE";
    PROPERTY["GRADIENT_STOPS"] = "GRADIENT_STOPS";
    PROPERTY["HOVERED"] = "HOVERED";
    PROPERTY["INTERPOLATION_POINTS"] = "INTERPOLATION_POINTS";
    PROPERTY["IS_DIGITAL_LINE"] = "IS_DIGITAL_LINE";
    PROPERTY["IS_ONE_HUNDRED_PERCENT"] = "IS_ONE_HUNDRED_PERCENT";
    PROPERTY["IS_SELECTED"] = "IS_SELECTED";
    PROPERTY["IS_VISIBLE"] = "IS_VISIBLE";
    PROPERTY["INTENSITY"] = "INTENSITY";
    PROPERTY["LINE_TYPE"] = "LINE_TYPE";
    PROPERTY["MAXIMUM"] = "MAXIMUM";
    PROPERTY["MINIMUM"] = "MINIMUM";
    PROPERTY["OFFSET"] = "OFFSET";
    PROPERTY["OPACITY"] = "OPACITY";
    PROPERTY["PALETTE_PROVIDER"] = "PALETTE_PROVIDER";
    PROPERTY["PARENT_SURFACE"] = "PARENT_SURFACE";
    PROPERTY["POINT_MARKER"] = "pointMarker";
    PROPERTY["RANGE"] = "RANGE";
    PROPERTY["RESAMPLING_MODE"] = "RESAMPLING_MODE";
    PROPERTY["RESAMPLING_PRECISION"] = "RESAMPLING_PRECISION";
    PROPERTY["SPACING"] = "SPACING";
    PROPERTY["STACKED_GROUP_ID"] = "STACKED_GROUP_ID";
    PROPERTY["STROKE"] = "STROKE";
    PROPERTY["STROKE_Y1"] = "STROKE_Y1";
    PROPERTY["STROKE_DASH_ARRAY"] = "STROKE_DASH_ARRAY";
    PROPERTY["STROKE_Y1_DASH_ARRAY"] = "STROKE_Y1_DASH_ARRAY";
    PROPERTY["STROKE_DOWN"] = "STROKE_DOWN";
    PROPERTY["STROKE_UP"] = "STROKE_UP";
    PROPERTY["STROKE_THICKNESS"] = "STROKE_THICKNESS";
    PROPERTY["USE_LINEAR_TEXTURE_FILTERING"] = "USE_LINEAR_TEXTURE_FILTERING";
    PROPERTY["XAXIS_ID"] = "XAXIS_ID";
    PROPERTY["YAXIS_ID"] = "YAXIS_ID";
    PROPERTY["ZERO_LINE_Y"] = "ZERO_LINE_Y";
    PROPERTY["Z_STEP"] = "Z_STEP";
    PROPERTY["Z_MAX"] = "Z_MAX";
    PROPERTY["Z_MIN"] = "Z_MIN";
    PROPERTY["Z_MULTIPLIER"] = "Z_MULTIPLIER";
    PROPERTY["Z_LABEL_PROVIDER"] = "Z_LABEL_PROVIDER";
    PROPERTY["SERIES_TEXT_PROVIDER"] = "SERIES_TEXT_PROVIDER";
    PROPERTY["CLIPTOYRANGE"] = "CLIPTOYRANGE";
    PROPERTY["YRANGEMODE"] = "YRANGEMODE";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
