// scichart/Charting/Visuals/Annotations/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DISTANCE_TO_LINE = exports.ADORNER_GRIP_RADIUS = exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["ADORNER_TEMPLATE"] = "ADORNER_TEMPLATE";
    PROPERTY["ANNOTATION_CANVAS"] = "ANNOTATION_CANVAS";
    PROPERTY["RESIZE_DIRECTIONS"] = "RESIZE_DIRECTIONS";
    PROPERTY["AXIS_LABELS_FILL"] = "AXIS_LABELS_FILL";
    PROPERTY["AXIS_LABELS_STROKE"] = "AXIS_LABELS_STROKE";
    PROPERTY["AXIS_FONT_FAMILY"] = "AXIS_FONT_FAMILY";
    PROPERTY["AXIS_FONT_SIZE"] = "AXIS_FONT_SIZE";
    PROPERTY["BACKGROUND_COLOR"] = "BACKGROUND_COLOR";
    PROPERTY["CLASS_NAME"] = "CLASS_NAME";
    PROPERTY["COLOR"] = "COLOR";
    PROPERTY["CONTAINER_BACKGROUND"] = "CONTAINER_BACKGROUND";
    PROPERTY["FILL"] = "FILL";
    PROPERTY["FONT_FAMILY"] = "FONT_FAMILY";
    PROPERTY["FONT_SIZE"] = "FONT_SIZE";
    PROPERTY["FONT_STYLE"] = "FONT_STYLE";
    PROPERTY["FONT_WEIGHT"] = "FONT_WEIGHT";
    PROPERTY["FORMATTED_VALUE"] = "FORMATTED_VALUE";
    PROPERTY["HORIZONTAL_ANCHOR_POINT"] = "HORIZONTAL_ANCHOR_POINT";
    PROPERTY["IMAGE"] = "IMAGE";
    PROPERTY["IMAGE_HEIGHT"] = "IMAGE_HEIGHT";
    PROPERTY["IMAGE_WIDTH"] = "IMAGE_WIDTH";
    PROPERTY["IS_EDITABLE"] = "IS_EDITABLE";
    PROPERTY["IS_HIDDEN"] = "IS_HIDDEN";
    PROPERTY["IS_SELECTED"] = "IS_SELECTED";
    PROPERTY["IS_HOVERED"] = "IS_HOVERED";
    PROPERTY["LABEL_PLACEMENT"] = "LABEL_PLACEMENT";
    PROPERTY["LABEL_VALUE"] = "LABEL_VALUE";
    PROPERTY["MINIMUM_WIDTH"] = "MINIMUM_WIDTH";
    PROPERTY["MINIMUM_HEIGHT"] = "MINIMUM_HEIGHT";
    PROPERTY["OPACITY"] = "OPACITY";
    PROPERTY["PADDING"] = "PADDING";
    PROPERTY["PARENT_SURFACE"] = "PARENT_SURFACE";
    PROPERTY["SERIES_INFO"] = "SERIES_INFO";
    PROPERTY["SERIES_INFOS"] = "SERIES_INFOS";
    PROPERTY["STROKE"] = "STROKE";
    PROPERTY["STROKE_THICKNESS"] = "STROKE_THICKNESS";
    PROPERTY["STROKE_DASH_ARRAY"] = "STROKE_DASH_ARRAY";
    PROPERTY["SHOW_LABEL"] = "SHOW_LABEL";
    PROPERTY["TEXT"] = "TEXT";
    PROPERTY["TEXT_STROKE"] = "TEXT_STROKE";
    PROPERTY["TITLE"] = "TITLE";
    PROPERTY["TOOLTIP_LEGEND_OFFSET_X"] = "TOOLTIP_LEGEND_OFFSET_X";
    PROPERTY["TOOLTIP_LEGEND_OFFSET_Y"] = "TOOLTIP_LEGEND_OFFSET_Y";
    PROPERTY["TOOLTIP_LEGEND_TEMPLATE"] = "TOOLTIP_LEGEND_TEMPLATE";
    PROPERTY["TOOLTIP_DATA_TEMPLATE"] = "TOOLTIP_DATA_TEMPLATE";
    PROPERTY["TOOLTIP_SVG_TEMPLATE"] = "TOOLTIP_SVG_TEMPLATE";
    PROPERTY["VERTICAL_ANCHOR_POINT"] = "VERTICAL_ANCHOR_POINT";
    PROPERTY["X1"] = "X1";
    PROPERTY["X2"] = "X2";
    PROPERTY["XAXIS_ID"] = "XAXIS_ID";
    PROPERTY["X_COORDINATE_MODE"] = "X_COORDINATE_MODE";
    PROPERTY["X_COORD_SHIFT"] = "X_COORD_SHIFT";
    PROPERTY["Y1"] = "Y1";
    PROPERTY["Y2"] = "Y2";
    PROPERTY["YAXIS_ID"] = "YAXIS_ID";
    PROPERTY["Y_COORDINATE_MODE"] = "Y_COORDINATE_MODE";
    PROPERTY["Y_COORD_SHIFT"] = "Y_COORD_SHIFT";
    PROPERTY["PLACEMENT_DIV_ID"] = "PLACEMENT_DIV_ID";
    PROPERTY["ROTATION"] = "ROTATION";
    PROPERTY["ADORNER_STROKE"] = "ADORNER_STROKE";
    PROPERTY["ADORNER_FILL"] = "ADORNER_FILL";
    PROPERTY["ADORNER_RADIUS"] = "ADORNER_RADIUS";
    PROPERTY["SELECTION_STROKE"] = "SELECTION_STROKE";
    PROPERTY["SELECTION_DELTA"] = "SELECTION_DELTA";
    PROPERTY["SELECTION_THICKNESS"] = "SELECTION_THICKNESS";
    PROPERTY["DRAGPOINTS"] = "DRAGPOINTS";
    PROPERTY["MULTILINE_ALIGNMENT"] = "MULTILINE_ALIGNMENT";
    PROPERTY["WRAP_TO"] = "WRAP_TO";
    PROPERTY["LINE_SPACING"] = "LINE_SPACING";
    PROPERTY["SCALE"] = "SCALE";
    PROPERTY["OUTLINE"] = "OUTLINE";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
exports.ADORNER_GRIP_RADIUS = 10;
exports.DISTANCE_TO_LINE = 10;
