// scichart/types/ChartModifierType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EChart3DModifierType = exports.EChart2DModifierType = void 0;
var EChart2DModifierType;
(function (EChart2DModifierType) {
    EChart2DModifierType["Cursor"] = "Cursor";
    EChart2DModifierType["DataPointSelection"] = "DataPointSelection";
    EChart2DModifierType["Legend"] = "Legend";
    EChart2DModifierType["MouseWheelZoom"] = "MouseWheelZoom";
    EChart2DModifierType["PinchZoom"] = "PinchZoom";
    EChart2DModifierType["Rollover"] = "Rollover";
    EChart2DModifierType["VerticalSlice"] = "VerticalSlice";
    EChart2DModifierType["RubberBandXYZoom"] = "RubberBandXYZoom";
    EChart2DModifierType["SeriesSelection"] = "SeriesSelection";
    EChart2DModifierType["XAxisDrag"] = "XAxisDrag";
    EChart2DModifierType["YAxisDrag"] = "YAxisDrag";
    EChart2DModifierType["ZoomExtents"] = "ZoomExtents";
    EChart2DModifierType["ZoomPan"] = "ZoomPan";
    EChart2DModifierType["OverviewRangeSelection"] = "OverviewRangeSelection";
    EChart2DModifierType["AnnotationHover"] = "AnnotationHover";
    EChart2DModifierType["Custom"] = "Custom";
})(EChart2DModifierType = exports.EChart2DModifierType || (exports.EChart2DModifierType = {}));
var EChart3DModifierType;
(function (EChart3DModifierType) {
    EChart3DModifierType["MouseWheelZoom"] = "MouseWheelZoom";
    EChart3DModifierType["Orbit"] = "Orbit";
    EChart3DModifierType["ZoomExtents"] = "ZoomExtents";
    EChart3DModifierType["Tooltip"] = "Tooltip";
    EChart3DModifierType["PinchZoom"] = "PinchZoom";
    EChart3DModifierType["Custom"] = "Custom";
})(EChart3DModifierType = exports.EChart3DModifierType || (exports.EChart3DModifierType = {}));
