// scichart/types/DefaultRenderLayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EDefaultRenderLayer = void 0;
var EDefaultRenderLayer;
(function (EDefaultRenderLayer) {
    EDefaultRenderLayer[EDefaultRenderLayer["Background"] = 0] = "Background";
    EDefaultRenderLayer[EDefaultRenderLayer["AxisBandsLayer"] = 1] = "AxisBandsLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["MinorGridLinesLayer"] = 2] = "MinorGridLinesLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["MajorGridLinesLayer"] = 3] = "MajorGridLinesLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["AnnotationsBelowSeriesLayer"] = 4] = "AnnotationsBelowSeriesLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["SeriesLayer"] = 5] = "SeriesLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["AxisBordersLayer"] = 6] = "AxisBordersLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["AxesLayer"] = 7] = "AxesLayer";
    EDefaultRenderLayer[EDefaultRenderLayer["AnnotationsAboveSeriesLayer"] = 8] = "AnnotationsAboveSeriesLayer";
})(EDefaultRenderLayer = exports.EDefaultRenderLayer || (exports.EDefaultRenderLayer = {}));
