// scichart/types/LabelPlacement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVerticalAlignment = exports.EHorizontalAlignment = exports.ELabelPlacement = void 0;
var ELabelPlacement;
(function (ELabelPlacement) {
    ELabelPlacement["Right"] = "Right";
    ELabelPlacement["TopRight"] = "TopRight";
    ELabelPlacement["BottomRight"] = "BottomRight";
    ELabelPlacement["Bottom"] = "Bottom";
    ELabelPlacement["Left"] = "Left";
    ELabelPlacement["TopLeft"] = "TopLeft";
    ELabelPlacement["BottomLeft"] = "BottomLeft";
    ELabelPlacement["Top"] = "Top";
    ELabelPlacement["Axis"] = "Axis";
    ELabelPlacement["Auto"] = "Auto";
})(ELabelPlacement = exports.ELabelPlacement || (exports.ELabelPlacement = {}));
var EHorizontalAlignment;
(function (EHorizontalAlignment) {
    EHorizontalAlignment["Right"] = "Right";
    EHorizontalAlignment["Left"] = "Left";
})(EHorizontalAlignment = exports.EHorizontalAlignment || (exports.EHorizontalAlignment = {}));
var EVerticalAlignment;
(function (EVerticalAlignment) {
    EVerticalAlignment["Top"] = "Top";
    EVerticalAlignment["Bottom"] = "Bottom";
})(EVerticalAlignment = exports.EVerticalAlignment || (exports.EVerticalAlignment = {}));
