// scichart/Charting/ChartModifiers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["ALLOWS_MULTI_SELECTION"] = "ALLOWS_MULTI_SELECTION";
    PROPERTY["ALLOW_TOOLTIP_OVERLAPPING"] = "ALLOW_TOOLTIP_OVERLAPPING";
    PROPERTY["CROSSHAIR_STROKE"] = "CROSSHAIR_STROKE";
    PROPERTY["EXECUTE_ON"] = "EXECUTE_ON";
    PROPERTY["FILL"] = "FILL";
    PROPERTY["HIT_TEST_RADIUS"] = "HIT_TEST_RADIUS";
    PROPERTY["IS_ATTACHED"] = "IS_ATTACHED";
    PROPERTY["IS_ENABLED"] = "IS_ENABLED";
    PROPERTY["PARENT_SURFACE"] = "PARENT_SURFACE";
    PROPERTY["RECEIVE_HANDLED_EVENTS"] = "RECEIVE_HANDLED_EVENTS";
    PROPERTY["SHOW_ROLLOVER_LINE"] = "SHOW_ROLLOVER_LINE";
    PROPERTY["SHOW_TOOLTIP"] = "SHOW_TOOLTIP";
    PROPERTY["SELECTED_AREA"] = "SELECTED_AREA";
    PROPERTY["SELECTION_FILL"] = "SELECTION_FILL";
    PROPERTY["SELECTION_STROKE"] = "SELECTION_STROKE";
    PROPERTY["SELECTION_STROKE_THICKNESS"] = "SELECTION_STROKE_THICKNESS";
    PROPERTY["STROKE"] = "STROKE";
    PROPERTY["STROKE_THICKNESS"] = "STROKE_THICKNESS";
    PROPERTY["STROKE_DASH_ARRAY"] = "STROKE_DASH_ARRAY";
    PROPERTY["TOOLTIP_DATA_TEMPLATE"] = "TOOLTIP_DATA_TEMPLATE";
    PROPERTY["X_AXIS_ID"] = "X_AXIS_ID";
    PROPERTY["Y_AXIS_ID"] = "Y_AXIS_ID";
    PROPERTY["PLACEMENT_DIV_ID"] = "PLACEMENT_DIV_ID";
    PROPERTY["X_LINE"] = "X_LINE";
    PROPERTY["Y_LINE"] = "Y_LINE";
    PROPERTY["AXIS_LABELS"] = "AXIS_LABELS";
    PROPERTY["X1"] = "X1";
    PROPERTY["X_COORDINATE_MODE"] = "X_COORDINATE_MODE";
    PROPERTY["IS_DRAGGABLE"] = "IS_DRAGGABLE";
    PROPERTY["LINE_ANNOTATIONS_OPTIONS"] = "LINE_ANNOTATIONS_OPTIONS";
    PROPERTY["LINE_SELECTION_COLOR"] = "LINE_SELECTION_COLOR";
    PROPERTY["TOOLTIP_SHADOW"] = "TOOLTIP_SHADOW";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
