// scichart/Charting/Visuals/Axis/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PROPERTY = void 0;
var PROPERTY;
(function (PROPERTY) {
    PROPERTY["ALLOW_FAST_MATH"] = "ALLOW_FAST_MATH";
    PROPERTY["AUTO_RANGE"] = "AUTO_RANGE";
    PROPERTY["AUTO_TICKS"] = "AUTO_TICKS";
    PROPERTY["AXIS_ALIGNMENT"] = "AXIS_ALIGNMENT";
    PROPERTY["AXIS_BANDS_FILL"] = "AXIS_BANDS_FILL";
    PROPERTY["AXIS_TITLE"] = "AXIS_TITLE";
    PROPERTY["AXIS_LENGTH"] = "AXIS_LENGTH";
    PROPERTY["BORDER"] = "BORDER";
    PROPERTY["CURSOR_TEXT_FORMATTING"] = "CURSOR_TEXT_FORMATTING";
    PROPERTY["DELTA_CALCULATOR"] = "DELTA_CALCULATOR";
    PROPERTY["TICK_COORDINATES_CALCULATOR"] = "TICK_COORDINATES_CALCULATOR";
    PROPERTY["DRAW_MAJOR_GRID_LINES"] = "DRAW_MAJOR_GRID_LINES";
    PROPERTY["DRAW_MINOR_GRID_LINES"] = "DRAW_MINOR_GRID_LINES";
    PROPERTY["DRAW_MAJOR_TICK_LINES"] = "DRAW_MAJOR_TICK_LINES";
    PROPERTY["DRAW_MINOR_TICK_LINES"] = "DRAW_MINOR_TICK_LINES";
    PROPERTY["DRAW_MAJOR_BANDS"] = "DRAW_MAJOR_BANDS";
    PROPERTY["DRAW_LABELS"] = "DRAW_LABELS";
    PROPERTY["FLIPPED_COORDINATES"] = "FLIPPED_COORDINATES";
    PROPERTY["GROW_BY"] = "GROW_BY";
    PROPERTY["ID"] = "ID";
    PROPERTY["IS_AXIS_FLIPPED"] = "IS_AXIS_FLIPPED";
    PROPERTY["IS_NEGATIVE"] = "IS_NEGATIVE";
    PROPERTY["IS_INNER_AXIS"] = "IS_INNER_AXIS";
    PROPERTY["IS_PRIMARY_AXIS"] = "IS_PRIMARY_AXIS";
    PROPERTY["IS_VISIBLE"] = "IS_VISIBLE";
    PROPERTY["IS_XAXIS"] = "IS_XAXIS";
    PROPERTY["LABEL_PROVIDER"] = "LABEL_PROVIDER";
    PROPERTY["AXIS_RENDERER"] = "LABEL_PROVIDER";
    PROPERTY["LOG_BASE"] = "LOG_BASE";
    PROPERTY["MAJOR_DELTA"] = "MAJOR_DELTA";
    PROPERTY["MAJOR_GRID_LINE_STYLE"] = "MAJOR_GRID_LINE_STYLE";
    PROPERTY["MAJOR_TICK_LINE_STYLE"] = "MAJOR_TICK_LINE_STYLE";
    PROPERTY["MAX_AUTO_TICKS"] = "MAX_AUTO_TICKS";
    PROPERTY["MINOR_DELTA"] = "MINOR_DELTA";
    PROPERTY["MINOR_GRID_LINE_STYLE"] = "MINOR_GRID_LINE_STYLE";
    PROPERTY["MINOR_TICK_LINE_STYLE"] = "MINOR_TICK_LINE_STYLE";
    PROPERTY["MINORS_PER_MAJOR"] = "MINORS_PER_MAJOR";
    PROPERTY["OFFSET"] = "OFFSET";
    PROPERTY["TEXT_FORMATTING"] = "TEXT_FORMATTING";
    PROPERTY["TEXT_STYLE"] = "TEXT_STYLE";
    PROPERTY["TICK_TEXT_BRUSH"] = "TICK_TEXT_BRUSH";
    PROPERTY["VISIBLE_RANGE"] = "VISIBLE_RANGE";
    PROPERTY["VISIBLE_RANGE_LIMIT"] = "VISIBLE_RANGE_LIMIT";
    PROPERTY["VISIBLE_RANGE_SIZE_LIMIT"] = "VISIBLE_RANGE_SIZE_LIMIT";
    PROPERTY["TICK_PROVIDER"] = "TICK_PROVIDER";
    PROPERTY["VIEW_RECT"] = "VIEW_RECT";
    PROPERTY["ZOOMEXTENTS_RANGE"] = "ZOOMEXTENTS_RANGE";
    PROPERTY["DRAW_ONLY_WITHIN_XRANGE"] = "DRAW_ONLY_WITHIN_XRANGE";
    PROPERTY["HIGH_PRECISION_TICKS"] = "HIGH_PRECISION_TICKS";
})(PROPERTY = exports.PROPERTY || (exports.PROPERTY = {}));
